### PR TITLE
Soup recipes that make items spawn the correct number of items

### LINKED ISSUE #77821: Cooked Noodles only produce 1 item at a time, but will consume all Fresh Noodles in the pot
## Round ID:
[213015](https://scrubby.melonmesa.com/round/213015)

## Testmerges:
[#77307: 'Revert "Bumps rust-g to `3.0.0` "' by MrStonedOne at commit 8dad0feceb](https://www.github.com/tgstation/tgstation/pull/77307)
[#77447: 'Trying to qdelete a callback now lags the server instead.' by MrStonedOne at commit 24565a3e08](https://www.github.com/tgstation/tgstation/pull/77447)

## Reproduction:
Put any number of Fresh Noodles in a soup pot with 30u of water and the required amount of salt and turn on the range. You will only receive one Cooked Noodle and all the extra Fresh Noodles in the pot as well as the all the extra salt will be consumed. You can waste a lot of ingredients if you don't make them one at a time.

### PR DESCRIPTION
**Pull request by blueDev2**
## About The Pull Request

Soup recipes, that make items, spawn the correct number of items per reaction instead of just one item
Soup recipes, that make and consume items, re-add unused reagents
Soup recipes that produce no reagents do not delete all required ingredients
## Why It's Good For The Game

Soup recipes should be correctly followed, however, there's some issues that happen with a soup recipe creates an item with no corresponding soup reagents.
1. Only 1 item is produced, no matter how many reagents or ingredients are used
2. All ingredients that are required_ingredients are deleted, this is done because the standard soup behavior would have transferred some of the reagents from any "required_ingredients into the soup. However, since no soup is made, the ingredients are wasted.
3. The reagents for the soup recipe itself are consumed to the highest multiple, even if there is insufficient  required_ingredients in the pot.

This PR ensure the proper number of items are spawned, the proper amount of required_reagents are used, and the proper amount of ingredients are used.

Fixes #77821
## Changelog
:cl:
fix: Soup recipes, that make items, spawn the correct number of items per reaction instead of just one item
fix: Soup recipes, that make items, consumes the correct number of reagents instead of the largest multiple of the reagents
/:cl:

### CODE REVIEW

File: code/modules/food_and_drinks/recipes/soup_mixtures.dm [hunk 1]
```diff
@@ -180,28 +207,25 @@
...
-		// Everything else will just get fried
-		else
-			ingredient.AddElement(/datum/element/fried_item, 30)
+			// Things that had reagents or ingredients in the soup will get deleted
+			else if(!isnull(ingredient.reagents) || is_type_in_list(ingredient, required_ingredients))
+				LAZYREMOVE(pot.added_ingredients, ingredient)
+				// Send everything left behind
+				transfer_ingredient_reagents(ingredient, holder)
+				// Delete, it's done
+				qdel(ingredient)
```
Comment (Jacquerel) [inline code/modules/food_and_drinks/recipes/soup_mixtures.dm:224]:
```suggestion
			if(isnull(ingredient.reagents) && !is_type_in_list(ingredient, required_ingredients))
				ingredient.AddElement(/datum/element/fried_item, 30)
				continue
			LAZYREMOVE(pot.added_ingredients, ingredient)
			// Send everything left behind
			transfer_ingredient_reagents(ingredient, holder)
			// Delete, it's done
			qdel(ingredient)
```

1. Don't need an else following continue
2. Let's get the shorter case out of the way first and return early

File: code/modules/food_and_drinks/recipes/soup_mixtures.dm [hunk 2]
```diff
@@ -85,6 +85,34 @@
 	if(!length(required_ingredients))
 		return
 
+	// If a food item is supposed to be made, remove relevant ingredients from the pot, then make the item 
+	if(resulting_food_path)
```
Comment (Jacquerel) [inline code/modules/food_and_drinks/recipes/soup_mixtures.dm:89]:
```suggestion
	if(!isnull(resulting_food_path))
```
In the spirit of san who is perpetually telling me to do this, let's be specific.

**Review by Jacquerel — CHANGES_REQUESTED** [review]
Couple of nitpicks

File: code/modules/food_and_drinks/recipes/soup_mixtures.dm (hunk 2 above)
Comment (blueDev2) [reply code/modules/food_and_drinks/recipes/soup_mixtures.dm:89]: Well, I guess that's fine. If this is something we should be doing, perhaps it should be included in the code style guide.

File: code/modules/food_and_drinks/recipes/soup_mixtures.dm (hunk 1 above)
Comment (Jacquerel) [reply code/modules/food_and_drinks/recipes/soup_mixtures.dm:224]: yeah i was too lazy to switch to side-to-side im afraid, you'd have to make the change manually 😅

Comment (blueDev2) [reply code/modules/food_and_drinks/recipes/soup_mixtures.dm]: Made changes, if your good with it, you can resolve this

**Pull request merged**